### ISSUE TITLE
SemidirectProduct Embedding does not return the proper subgroup?

### ISSUE THREAD

**Issue by hungaborhorvath** [gap-system/gap#565]
According to the manual, ````SemidirectProduct```` ````Embedding```` should return the embeddings of the normal subgroup and the complement. This is indeed the case for the following semidirect product of S_4, and the embedding does not give the whole group:
````
gap> N := Group((1,2)(3,4), (1,3)(2,4));; H := AutomorphismGroup(N);;
gap> G := SemidirectProduct(H, N);;
gap> Size(Image(Embedding(G, 2)))=Size(N);
true
gap> Size(Image(Embedding(G, 1)))=Size(H);
true
gap> Image(Embedding(G, 1))=G;
false
````
However, for the semidirect product of A_6 with its automorphism group ````Embedding```` 1 gives the whole group:
````
gap> N := AlternatingGroup(6);; H := AutomorphismGroup(N);;
gap> G := SemidirectProduct(H, N);;
gap> Size(Image(Embedding(G, 2)))=Size(N);
true
gap> Size(Image(Embedding(G, 1)))=Size(H);
false
gap> Image(Embedding(G, 1))=G;
true
````
Apparently, up to A_20, there is no problem, except for A_6.
Am I missing something trivial here or is this a bug?

**Comment by hulpke** [comment]
This is a bug, and its surprising it only comes up now. As part of testing for a map being a homomorphism, an inverse general mapping (defined only on the image) is computed.

There is an immediate method in ghom.gi (this has been a pain to debug...) that set `IsSurjective` if an inverse general mapping exists and that is total. However the general convention of (at least permutation) homomorphisms is that the inverse general mapping is not necessarily defined on the range, but on the image. So this immediate method should not apply here.

While one could try to fix this by adding a test for the domain of the inverse being equal to the range, this makes the method more messy and takes time.

I therefore propose to remove this immediate methods (and its relatives) -- the code for inverse mappings in fact explicitly contains `set' commands so that these immediate methods are unlikely to be of use any longer.

**Comment by fingolfin** [comment]
Actually, to me the real problem seems to be that there are general mapping which violate the contract "Range(map) = Source(Inverse(map))"

If we abolish that, we will have to remove a lot more than just those immediate methods, I think. A lot of the mapping code is based on this assumption (IIRC -- it's been some years, but I once dug very thoroughly through this code, in order to make the homomorphisms of the polycylic and polenta packages work correctly).

**Comment by hungaborhorvath** [comment]
Just our of curiosity, how come this only happens for A6, and not for the other alternating groups? I am sure it has something to do with the fact that its AutomorphismGroup is not just S6, but S6:C2, and I saw that the permutation representation of the semidirect product uses action on 360 elements (I am guessing the elements of A6), I just do not really understand why this makes such a mess for GAP....

**Comment by hulpke** [comment]
@hungaborhorvath For An different to 6 the automorphisms cannot be represented in the same representation as An, thus a different code path runs.

**Comment by fingolfin** [comment]
For the record, this bug is already present in GAP 4.4.12 and 4.5.3

**Comment by hulpke** [comment]
Danke. H”atte ich beim das mergen schon machen sollen (oder ein `close’ in den committ text schreiben?) 

Alexander

**Comment by fingolfin** [comment]
@hulpke Actually, *you* ensured that this was closed :-). Namely, by writing the text: "fixes #NNN" into the commit message. This is the same as "closes #NNN", and various others, see https://help.github.com/articles/closing-issues-via-commit-messages/

**Comment by hulpke** [comment]
> On Feb 17, 2016, at 9:32 AM, Max Horn <notifications@github.com> wrote:
> 
> @hulpke Actually, you ensured that this was closed :-). Namely, by writing the text: "fixes #NNN" into the commit message. This is the same as "closes #NNN", and various others, see https://help.github.com/articles/closing-issues-via-commit-messages/

Thank you. That means the message

(Markus Pfeiffer:  Closed #565 via 703be04.   )

was produced automatically?

Thanks,

   Alexander

**Comment by fingolfin** [comment]
Exactly.
